[UI] Volumes Sorting

alexis-ld opened the pull request:
**Component**:
ui, volumes

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 
We want to add sorting to the Volumes list.

**Summary**:
I implemented sorting capacities using react-table's custom sorting functions. I also pushed the sorting parameters in the URL as required in #2876 . If a sorting is set in the URL when loading the page it is applied if not is default back to health sorting.

**Acceptance criteria**: 
Clicking on a column name sorts the list based on the selected parameter.
The sorting for each columns follows the specifications described in #2876 .
If a sort( and optional desc) parameter i set in the URL it is applied on page load.

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2876 

<!-- If you want to refer to an issue while not closing it, use:


-->

Commit:
ui/volumes:
Adapt getVolumeListData for table sorting. Removed initial sorting and export latency as a numeric value.

Refs: #2876

Commit:
ui/volumes:
Implement react-table custom sorting for volume list.

Refs: #2876

Commit:
ui/volumes:
Set react-table's autoResetSortBy to false to keep sorting active when data refreshes.

Refs: #2876

Commit:
ui/volumes:
Store volumes sort parameters in URL.
Apply sorting if specified in URL.

Refs: #2876

Commit:
ui/volumes:
Extract health comparison logic to utils file to use it both on initial data sorting and react-table sorting.

Refs: #2876

Commit:
ui/volumes:
Remove URL param query when sorting resetted back to default.

Refs: #2876

Commit:
ui/volumes: Adapt getVolumeListData tests for new data formatting.

Refs: #2876

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2876-volumes-sorting` into
`development/2.6`
* [w/2.7/improvement/2876-volumes-sorting](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2876-volumes-sorting) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

thomasdanan reviewed (commented):
Works well.
Just noticed that some columns header are centered and some other are aligned on the left. When it is centered, activating the makes the header to move on the left, which is not great. Shouldn't we have all headers aligned on the left so that when we activate sorting, headers are not moving on the left?

Also, it is not super obvious that one can click on the header to sort. Could we make the sorting icon to appear when we mouse over the column header?

Since there is a default sorting (on Health), shouldn't we display the fact that the table is sorted on health when loading on the page?

Commit:
ui/volumes:
Add an absolute positionning on caret to avoid it moving the label when sorting.

Commit:
ui/volumes:
Display a sort incentive when the user hovers the column header

alexis-ld commented:
@thomasdanan Just added a fix for the header moving on sort (I managed to keep their initial positioning but still not move them when the sorting caret appears). I also added the sort incentive on hover.

Working on the initial sorting caret...

thomasdanan commented:
> @thomasdanan Just added a fix for the header moving on sort (I managed to keep their initial positioning but still not move them when the sorting caret appears). I also added the sort incentive on hover.
> 
> Working on the initial sorting caret...

Thanks @alexis-ld, actually we had a discussion with @ChengYanJin and with our UX designer. We were saying that for table values, icon values should be center aligned, string values should be left-aligned and numerical values should be right-aligned. Should this rule apply to the header or should we have headers always left-aligned? we were thinking having header left-aligned would make it easier for you to fix the moving header issue.

alexis-ld commented:
@thomasdanan for now the code applies the same alignement on the header as the value. It looks a bit weird to have a centered value with a left header actually (the offset is quite important).

Anyway I fixed the moving header thing without moving the header to the left so we can keep it like that if that works for you.

Commit:
ui/volumes:
Enables default health sorting directly within react-table (to display caret)

alexis-ld commented:
I found a way to use react-table's default sorting without broking the URL sync. Should be all good now :) .

ChengYanJin commented on ui/src/components/VolumeListTable.js:
```diff
@@ -115,6 +124,25 @@ const TooltipContent = styled.div`
+`;
+
+const SortIncentive = styled.span`
+  padding-left: ${padding.smaller};
```
I feel it's better to remove the padding-left, so when we click on the sort the caret would not jump.

ChengYanJin reviewed (changes requested):
It works!! 

I still encounter the infinite loop of URL by the following steps:
Sort by other columns => click on "Create Volume" => click on "Cancel" button. 

Can we try to set the initial sort in React-table, instead of initialing the data sorting in NodeVolumesUtils.js?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

Commit:
ui/volumes:
Fix URL loop again by using react-table initial sort with URL queries.

Refs: #2876

Commit:
ui/volumes: Removed padding on sortIncentive for better alignment

ChengYanJin reviewed (approved):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexis-ld commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexis-ld.
